packages/kokkos-kernels/src/impl/tpls/KokkosSparse_spmv_tpl_spec_decl.hpp

```diff
@@ -209,33 +250,43 @@ namespace Impl {
-  // KOKKOSSPARSE_SPMV_CUSPARSE(double, int64_t, int, Kokkos::LayoutLeft,  false)
-  KOKKOSSPARSE_SPMV_CUSPARSE(double, int64_t, int, Kokkos::LayoutRight, true)
-  // KOKKOSSPARSE_SPMV_CUSPARSE(double, int64_t, int, Kokkos::LayoutRight, false)
-  KOKKOSSPARSE_SPMV_CUSPARSE(float,  int64_t, int, Kokkos::LayoutLeft,  true)
-  // KOKKOSSPARSE_SPMV_CUSPARSE(float,  int64_t, int, Kokkos::LayoutLeft,  false)
-  KOKKOSSPARSE_SPMV_CUSPARSE(float,  int64_t, int, Kokkos::LayoutRight, true)
-  // KOKKOSSPARSE_SPMV_CUSPARSE(float,  int64_t, int, Kokkos::LayoutRight, false)
-  KOKKOSSPARSE_SPMV_CUSPARSE(double, int64_t, int64_t, Kokkos::LayoutLeft,  true)
-  // KOKKOSSPARSE_SPMV_CUSPARSE(double, int64_t, int64_t, Kokkos::LayoutLeft,  false)
-  KOKKOSSPARSE_SPMV_CUSPARSE(double, int64_t, int64_t, Kokkos::LayoutRight, true)
-  // KOKKOSSPARSE_SPMV_CUSPARSE(double, int64_t, int64_t, Kokkos::LayoutRight, false)
-  KOKKOSSPARSE_SPMV_CUSPARSE(float,  int64_t, int64_t, Kokkos::LayoutLeft,  true)
-  // KOKKOSSPARSE_SPMV_CUSPARSE(float,  int64_t, int64_t, Kokkos::LayoutLeft,  false)
-  KOKKOSSPARSE_SPMV_CUSPARSE(float,  int64_t, int64_t, Kokkos::LayoutRight, true)
-  // KOKKOSSPARSE_SPMV_CUSPARSE(float,  int64_t, int64_t, Kokkos::LayoutRight, false)
+  }; 
+#if (9000 <= CUSPARSE_VERSION)
+  KOKKOSSPARSE_SPMV_CUSPARSE(double, int, int, Kokkos::LayoutLeft,  Kokkos::CudaSpace, true)
+  KOKKOSSPARSE_SPMV_CUSPARSE(double, int, int, Kokkos::LayoutRight, Kokkos::CudaSpace, true)
+  KOKKOSSPARSE_SPMV_CUSPARSE(float,  int, int, Kokkos::LayoutLeft,  Kokkos::CudaSpace, true)
+  KOKKOSSPARSE_SPMV_CUSPARSE(float,  int, int, Kokkos::LayoutRight, Kokkos::CudaSpace, true)
+  KOKKOSSPARSE_SPMV_CUSPARSE(Kokkos::complex<double>, int, int, Kokkos::LayoutLeft,  Kokkos::CudaSpace, true)
+  KOKKOSSPARSE_SPMV_CUSPARSE(Kokkos::complex<double>, int, int, Kokkos::LayoutRight, Kokkos::CudaSpace, true)
+  KOKKOSSPARSE_SPMV_CUSPARSE(Kokkos::complex<float>,  int, int, Kokkos::LayoutLeft,  Kokkos::CudaSpace, true)
+  KOKKOSSPARSE_SPMV_CUSPARSE(Kokkos::complex<float>,  int, int, Kokkos::LayoutRight, Kokkos::CudaSpace, true)
+  KOKKOSSPARSE_SPMV_CUSPARSE(double, int, int, Kokkos::LayoutLeft,  Kokkos::CudaUVMSpace, true)
+  KOKKOSSPARSE_SPMV_CUSPARSE(double, int, int, Kokkos::LayoutRight, Kokkos::CudaUVMSpace, true)
+  KOKKOSSPARSE_SPMV_CUSPARSE(float,  int, int, Kokkos::LayoutLeft,  Kokkos::CudaUVMSpace, true)
+  KOKKOSSPARSE_SPMV_CUSPARSE(float,  int, int, Kokkos::LayoutRight, Kokkos::CudaUVMSpace, true)
+  KOKKOSSPARSE_SPMV_CUSPARSE(Kokkos::complex<double>, int, int, Kokkos::LayoutLeft,  Kokkos::CudaUVMSpace, true)
+  KOKKOSSPARSE_SPMV_CUSPARSE(Kokkos::complex<double>, int, int, Kokkos::LayoutRight, Kokkos::CudaUVMSpace, true)
+  KOKKOSSPARSE_SPMV_CUSPARSE(Kokkos::complex<float>,  int, int, Kokkos::LayoutLeft,  Kokkos::CudaUVMSpace, true)
+  KOKKOSSPARSE_SPMV_CUSPARSE(Kokkos::complex<float>,  int, int, Kokkos::LayoutRight, Kokkos::CudaUVMSpace, true)
+
+#if (10300 <= CUSPARSE_VERSION)
```

Comment:
@lucbv I did try setting it to COMPILE_LIBRARY, but then I got undefined references in tests. All the BLAS TPL spec_decl.hpp files set it to true, and it makes the build succeed, so I guess that's what it should be. Maybe @jjwilke understands this better.

Reply:
@lucbv Yeah, I think that it should be true. I dumped the symbols out of all the Tpetra object files, since several of them include "KokkosSparse_spmv.hpp" but none of them should contain weak symbols of ``Impl::SPMV<...>::spmv`` for types that KK explicitly instantiates. And they don't, all of those are undefined. So I think we're good.